internal/firestoretest/emulator.go

```diff
@@ -0,0 +1,103 @@
+package firestoretest
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"net/http"
+	"os"
+	"os/exec"
+	"testing"
+	"time"
+
+	"github.com/pkg/errors"
+)
+
+func getPort() (*int, error) {
+	var localhost [net.IPv6len]byte
+	localhost[len(localhost)-1]++ // ::1
+	l, err := net.ListenTCP("tcp", &net.TCPAddr{IP: net.IP(localhost[:]), Port: 0})
+	if err != nil {
+		return nil, err
+	}
+	defer l.Close()
+	return &l.Addr().(*net.TCPAddr).Port, nil
+}
+
+func waitForCmd(cmd *exec.Cmd) <-chan struct{} {
+	cmdDone := make(chan struct{})
+	go func() {
+		cmd.Wait()
+		close(cmdDone)
+	}()
+	return cmdDone
+}
+
+func StartEmulator(ctx context.Context, t *testing.T) <-chan error {
+	t.Helper()
+	port, err := getPort()
+	if err != nil {
+		t.Fatalf("getPort(): %v", err)
+	}
+	addr := fmt.Sprintf("localhost:%d", *port)
+	t.Logf("starting firestore emulator... addr=%s", addr)
+
+	cmd := exec.Command("gcloud", "emulators", "firestore", "start", "--host-port="+addr)
+	cmd.Stderr = os.Stderr
+	cmd.Stdout = os.Stderr
+	if err := cmd.Start(); err != nil {
+		t.Fatalf("failure starting firestore emulator: %v", err)
+	}
+
+	if err := os.Setenv("FIRESTORE_EMULATOR_HOST", addr); err != nil {
+		t.Fatalf("os.SetEnv(): %v", err)
+	}
+
+	result := make(chan error, 1)
+	go func() {
+		portReachable := make(chan struct{})
+		go func() {
+			for {
+				c, err := net.DialTCP("tcp", nil, &net.TCPAddr{Port: *port})
+				if err == nil {
+					c.Close()
+					close(portReachable)
+					return
+				}
+				time.Sleep(300 * time.Millisecond)
+			}
```

Comment:
If firestore emulator fails to start, will this goroutine ever be cleaned up? Should this also attempt to wait for ctx.Done()?

Reply:
sure although in that case, the program's gonna exit immediately so i see very little value in it